Memory leak with director, csharp, std::string and sending strings to c#

## SWIG.i

```c++
%{
#include "net_invoker.h"
%}

%feature("director") NetInvokerBase;

class NetInvokerBase {
public:
    virtual ~NetInvokerBase() {}
    virtual const NetMethodInfo* GetMethodInfo(std::string tt)
    {
        return NULL;
    }
};

class NetInvoker {
public:
  static void set(NetInvokerBase* invoker) { NetInvoker::invoker = invoker; }
  static void reset() { NetInvoker::invoker = 0; }
private:
  static NetInvokerBase *invoker;
};
```

## C#

```c#
public class RuntimeNetInvoker : Qt.NetCore.NetInvokerBase
{
    public override bool IsValidType(string type)
    {
        return true;
    }

    public override NetMethodInfo GetMethodInfo(string tt)
    {
        return null;
    }
}

NetInvoker.set(new RuntimeNetInvoker());
```

## C++

```c++
std::string netType = "test";
NetInvoker::invoker->GetMethodInfo(netType);
```

This problem goes away if I take out the ```std::string``` parameter to ```GetMethodInfo```. So, this told me the issue is somewhere in the generated c++ code that marshals the ```std::string``` to the c# callback.

Here was the original code that was generated.

```c++
NetMethodInfo const *SwigDirector_NetInvokerBase::GetMethodInfo(std::string tt) {
  NetMethodInfo *c_result = 0 ;
  void * jresult = 0 ;
  char * jtt  ;
  
  if (!swig_callbackGetMethodInfo) {
    return NetInvokerBase::GetMethodInfo(tt);
  } else {
    jtt = SWIG_csharp_string_callback((&tt)->c_str());
    jresult = (void *) swig_callbackGetMethodInfo(jtt);
    c_result = (NetMethodInfo *)jresult; 
  }
  return (NetMethodInfo const *)c_result;
}
```

If I take out the ```SWIG_csharp_string_callback``` call, and passed the string directly to ```swig_callbackGetMethodInfo```, everything works, and the memory leak is gone.

So, changing the generated wrapper to this, resolved it.

```c#
NetMethodInfo const *SwigDirector_NetInvokerBase::GetMethodInfo(std::string tt) {
  NetMethodInfo *c_result = 0 ;
  void * jresult = 0 ;
  char * jtt  ;
  
  if (!swig_callbackGetMethodInfo) {
    return NetInvokerBase::GetMethodInfo(tt);
  } else {
    //jtt = SWIG_csharp_string_callback((&tt)->c_str());
    jresult = (void *) swig_callbackGetMethodInfo((&tt)->c_str());
    c_result = (NetMethodInfo *)jresult; 
  }
  return (NetMethodInfo const *)c_result;
}
```
How do I resolve this? I suspect the issue is somewhere in ```%include "std_string.i"```. for c#.

I am using swig 3.0.8, fresh from the Ubuntu repos.

I tried the patch here with the testsuite and there are const issues which all look very like this:

```
checking csharp testcase director_default
director_default_wrap.cxx: In member function ‘virtual std::string SwigDirector_Foo::Msg(std::string)’:
director_default_wrap.cxx:435:25: error: invalid conversion from ‘const char*’ to ‘char*’ [-fpermissive]
  435 |     jmsg = (&msg)->c_str();
      |            ~~~~~~~~~~~~~^~
      |                         |
      |                         const char*
make[2]: *** [../../../../Examples/Makefile:1130: csharp_cpp] Error 1
make[1]: *** [Makefile:61: director_default.cpptest] Error 2
```

The reproducer from #283 has the same issue.

But if I also change:

```
-%typemap(ctype) string "char *"
+%typemap(ctype) string "const char *"
```

that fixes #283.  There's a testsuite failure which seems unrelated.  I'll poke a bit and open a PR if I can get this to work.

--- a/CHANGES.current
+++ b/CHANGES.current
@@ -8,6 +8,10 @@ Version 4.1.0 (in progress)
 ===========================
 
 2022-02-03: olly
+	    [C#] #283 #998 Fix memory leak in directorin typemap for
+	    std::string.
+
+2022-02-03: olly
 	    [Python] #967 Make `self` parameter available to user typemaps.
 
 2022-02-03: teythoon

--- a/Lib/csharp/std_string.i
+++ b/Lib/csharp/std_string.i
@@ -20,7 +20,7 @@ namespace std {
 class string;
 
 // string
-%typemap(ctype) string "char *"
+%typemap(ctype) string "const char *"
 %typemap(imtype) string "string"
 %typemap(cstype) string "string"
 
@@ -42,7 +42,7 @@ class string;
    }
    $result.assign($input); %}
 
-%typemap(directorin) string %{ $input = SWIG_csharp_string_callback($1.c_str()); %}
+%typemap(directorin) string %{ $input = $1.c_str(); %}
 
 %typemap(csin) string "$csinput"
 %typemap(csout, excode=SWIGEXCODE) string {

